kong/plugins/http-log/handler.lua

```diff
@@ -18,68 +17,146 @@ HttpLogHandler.PRIORITY = 12
---  end
 
-  self.name = name
+-- Sends the provided payload (a string) to the configured plugin host
+-- @return true if everything was sent correctly, falsy if error
+-- @return error message if there was an error
+local function send_payload(self, conf, payload)
+  local method = conf.method
+  local timeout = conf.timeout
+  local keepalive = conf.keepalive
+  local content_type = conf.content_type
+  local http_endpoint = conf.http_endpoint
+
+  local ok, err
+  local parsed_url = parse_url(http_endpoint)
+  local host = parsed_url.host
+  local port = tonumber(parsed_url.port)
+
+  local httpc = http.new()
+  httpc:set_timeout(timeout)
+  ok, err = httpc:connect(host, port)
+  if not ok then
+    kong.log.err("failed to connect to ", host, ":", tostring(port), ": ", err)
+    return false
+  end
+
+  if parsed_url.scheme == "https" then
+    local _, err = httpc:ssl_handshake(true, host, false)
+    if err then
+      kong.log.err("failed to do SSL handshake with ",
+                   host, ":", tostring(port), ": ", err)
+      return false
+    end
+  end
+
+  local res, err = httpc:request({
+    method = method,
+    path = parsed_url.path,
+    query = parsed_url.query,
+    headers = {
+      ["Host"] = parsed_url.host,
+      ["Content-Type"] = content_type,
+      ["Content-Length"] = #payload,
+      ["Authorization"] = parsed_url.userinfo and (
+        "Basic " .. ngx_encode_base64(parsed_url.userinfo)
+      ),
+    },
+    body = payload,
+  })
+  if not res then
+    kong.log.err("failed request to ", host, ":", tostring(port), ": ", err)
+    httpc:set_keepalive(keepalive)
+    return false
+  end
+
+  -- read and discard response body
+  -- TODO should we fail if response status was >= 500 ?
+  res:read_body()
+
+  ok, err = httpc:set_keepalive(keepalive)
+  if not ok then
+    kong.log.err("failed keepalive for ", host, ":", tostring(port), ": ", err)
+  end
+
+  return true
 end
 
 
--- serializes context data into an html message body.
--- @param `ngx` The context table for the request being logged
--- @param `conf` plugin configuration table, holds http endpoint details
--- @return html body as string
-function HttpLogHandler:serialize(ngx, conf)
-  return cjson_encode(basic_serializer.serialize(ngx))
+local function json_array_concat(entries)
+  return "[" .. table.concat(entries, ",") .. "]"
+end
+
+
+-- Only provide `name` when deriving from this class,
+-- not when initializing an instance.
+function HttpLogHandler:new(name)
+  name = name or "http-log"
+  HttpLogHandler.super.new(self, name)
+
+  self.name = name
 end
 
 
 function HttpLogHandler:log(conf)
   HttpLogHandler.super.log(self)
 
-  local route_id = conf.route_id or "global"
-  local buf = buffers[route_id]
-  if not buf then
-
-    if conf.queue_size == nil then
-      conf.queue_size = 1
-    end
+  local entry = cjson_encode(basic_serializer.serialize(ngx))
 
+  local route_id = conf.route_id or "global"
+  local q = queues[route_id]
```

Comment:
caching it by route is weird? shouldn't it be cached by plugin `conf` table?

Reply:
Also `conf.route_id` (slightly above) is always `nil`. `route_id` lives on the top level. (it is passed on the management api as `route_id` and not as `config.route_id`)

Reply:
I understand why you might think that. You got trumped by the magic of the plugins iterator. `conf.route_id` it is populated by it here: https://github.com/Kong/kong/blob/master/kong/runloop/plugins_iterator.lua#L64

I also think it is a bit weird to do this by route instead of by other criteria, but that's how it was done before this refactor so I decided to leave that fight for another day.

> caching it by route is weird? shouldn't it be cached by plugin conf table?

I would cache it by plugin.id if I had it. I'm hesitant to use the conf table since I am not 100% sure those tables are never duplicated.

Reply:
since this uses the mlcache to store it I guess its shared with other workers and serialized?, hence using the `conf` table itself as a key doesn't work I guess.

I think what really makes sense here is to create 1 queue per log server-config. So maybe construct the cache key from that info?

Reply:
> I think what really makes sense here is to create 1 queue per log server-config. 

This makes sense, but there's a mismatch in that `queue_size` is defined per plugin. Agree with @kikito that `plugin_id` would be ideal if available (given current public interface for the plugin), but yeah, this should suffice for now.

Reply:
After talking privately with @Tieske and @hishamhm about this we decided to go with an alternative approach which uses "a serialized form of the plugin config" as a queue key instead of the `route_id` which was used previously.